### PR TITLE
Interactive login

### PR DESCRIPTION
**Pull request by supreeth7**
[OSD-8115](https://issues.redhat.com/browse/OSD-8115), [OSD-8135](https://issues.redhat.com/browse/OSD-8135), [OSD-8138](https://issues.redhat.com/browse/OSD-8138), [OSD-8195](https://issues.redhat.com/browse/OSD-8195)

### CODE REVIEW

**Comment by T0MASD** [comment]
I'm unable to run the tests: 
```
$ make test
Makefile:58: warning: overriding recipe for target 'test'
Makefile:54: warning: ignoring old recipe for target 'test'
which go-bindata || go get github.com/go-bindata/go-bindata/go-bindata
/usr/bin/which: no go-bindata in (/home/tomas/.local/bin:/home/tomas/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/bin)
go generate -x ./cmd/...
for cmd in $(ls cmd); do \
	go build -o "${cmd}" "./cmd/${cmd}" || exit 1; \
done
package test is not in GOROOT (/usr/lib/golang/src/test)
make: *** [Makefile:58: test] Error 1
```
I tried running make tools, but that didn't help either
```
$ make tools
Makefile:58: warning: overriding recipe for target 'test'
Makefile:54: warning: ignoring old recipe for target 'test'
which go-bindata || go get github.com/go-bindata/go-bindata/go-bindata
/usr/bin/which: no go-bindata in (/home/tomas/.local/bin:/home/tomas/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/bin)
```

**Comment by supreeth7** [comment]
@T0MASD fixed the makefile, tests haven't yet been implemented though so expect them to fail. Will be adding tests through another PR.

File: .gitignore [hunk 1]
```diff
@@ -12,7 +12,3 @@
 *.out
 
 # Dependency directories (remove the comment below to include it)
-# vendor/
-
-# PDCLI binary
-pdcli
```
Comment (chopraapooja) [inline .gitignore:18]: We should keep pdcli binary in .gitignore, we don't want to check in binary.

Comment (chopraapooja) [reply .gitignore:18]: @supreeth7  Did you do it intentionally? I think you should have run rebase while pulling changes from upstream.

File: cmd/pdcli/login/cmd.go [hunk 2]
```diff
@@ -13,31 +13,168 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
+	"os"
 
+	"github.com/PagerDuty/go-pagerduty"
+	"github.com/openshift/pagerduty-short-circuiter/pkg/config"
+	"github.com/openshift/pagerduty-short-circuiter/pkg/pdcli"
 	"github.com/spf13/cobra"
 )
 
-// pdcli/login/cmdCmd represents the pdcli/login/cmd command
+const APIKeyURL = "https://support.pagerduty.com/docs/generating-api-keys#section-generating-a-general-access-rest-api-key"
```
Comment (chopraapooja) [inline cmd/pdcli/login/cmd.go:31]: Let's take APIKeyURL into a separate package something like `constants` which will hold the constant data. How about `pkg/constants` ?

File: cmd/pdcli/login/cmd.go [hunk 3]
```diff
@@ -13,31 +13,168 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-	Short: "A brief description of your command",
-	Long: `A longer description that spans multiple lines and likely contains examples
-and usage of using your command. For example:
-`,
-	Run: func(cmd *cobra.Command, args []string) {
-		fmt.Println("logged in")
-	},
+	Short: "Login to the PagerDuty CLI",
+	Long: `Running the pdcli login command will send a request to PagerDuty REST API provided a valid API key.
+The PagerDuty REST API supports authenticating via the user API token.`,
```
Comment (chopraapooja) [inline cmd/pdcli/login/cmd.go:41]: Long:  We don't have to explain internal implementations to the end user, think about what more information, you want to tell the user, if there is none, you can leave this part.

File: pkg/config/config.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+// If the config filepath doesn't exist, the desired config filepath string is returned.
+func Find() (string, error) {
+	//locates the user home directory
+	homedir, err := os.UserHomeDir()
+
+	if err != nil {
+		return "", fmt.Errorf("cannot locate user home directory")
+	}
+
+	configPath := filepath.Join(homedir, ".config/pagerduty-cli/config.json")
```
Comment (chopraapooja) [inline pkg/config/config.go:44]: move `.config/pagerduty-cli/config.json` to constants pkg

File: .gitignore (hunk 1 above)
Comment (supreeth7) [reply .gitignore:18]: Yes Pooja, it was intentional. By adding 'pdcli' to .gitignore it stops tracking the entire cmd/pdcli folder. So I've tweaked the make file commands to create the binary directly in the ${GOPATH}/bin itself making it easier to natively run the pdcli command. For tests, the binary is created in the project root and deleted automatically upon completion.

File: cmd/pdcli/login/cmd.go [hunk 5]
```diff
@@ -13,31 +13,168 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
+// loginHandler handles the login flow into pdcli.
+func loginHandler(cmd *cobra.Command, args []string) error {
+
+	// load configuration info
+	cfg, err := config.Fetch()
+
+	// if the config file is located
+	// check if config file is empty, initialize a new config struct
+	if cfg == nil {
+		cfg = new(config.Config)
```
Comment (chopraapooja) [inline cmd/pdcli/login/cmd.go:60]: Isn't you are already raising error in Fetch for this? 
```
if len(configData) == 0 {
		err = fmt.Errorf("configuration file is empty")
		return nil, err
	}```

File: .gitignore (hunk 1 above)
Comment (chopraapooja) [reply .gitignore:18]: @supreeth7  try `./pdcli`

**Comment by chopraapooja** [comment]
@supreeth7 would you please create one Jira with all the pending items in this PR, I am merging this one.

File: .gitignore (hunk 1 above)
Comment (supreeth7) [reply .gitignore:18]: Nope, that doesn't work. :)

**Pull request merged**